nx/lib/nx/random.ex

```diff
@@ -0,0 +1,419 @@
+  an `Nx.Tensor` composed by 2 unsigned 32-bit integers
+  that we call a key, usually generated by the `Nx.Random.key/1`
+  function:
+  
+      iex> Nx.Random.key(12)
+      #Nx.Tensor<
+        u32[2]
+        [0, 12]
+      >
+  
+  This key can then be used in any of
+  Nx’s random number generation routines:
+  
+      iex> key = Nx.Random.key(12)
+      iex> Nx.Random.uniform(key)
+      #Nx.Tensor<
+        f32[1]
+        [0.3037703037261963]
+      >
+  
+  Note that using a key does not modify it, so
+  reusing the same key will lead to the same result:
+  
+      iex> key = Nx.Random.key(12)
+      iex> Nx.Random.uniform(key)
+      #Nx.Tensor<
+        f32[1]
+        [0.3037703037261963]
+      >
+  
+  If you need a new random number, you
+  can use `Nx.Random.split/2` to generate new subkeys:
+  
+      iex> key = Nx.Random.key(12)
+      iex> keys = Nx.Random.split(key)
+      iex> Nx.Random.uniform(keys[0])
+      #Nx.Tensor<
+        f32[1]
+        [0.1531379222869873]
+      >
+      iex> Nx.Random.uniform(keys[1])
+      #Nx.Tensor<
+        f32[1]
+        [0.7691127061843872]
+      >
+  
+  ## Advanced
+  
+  Design and Context
+  
+   In short, Nx's PRNGs are based on a Threefry counter PRNG
+   associated to a functional array-oriented splitting model
+  
+  To summarize, among other requirements, Nx's PRNG aims to:
+  
+  1. Ensure reproducibility;
+  
+  2. Parallelize well, both in terms of vectorization (generating array values)
+     and multi-replica, multi-core computation. In particular it should not use
+     sequencing constraints between random function calls.
+  """
+
+  import Nx.Defn, only: [deftransformp: 2, defn: 2, defnp: 2]
+
+  @nbits 32
+
+  @doc """
+  Create a pseudo-random number generator (PRNG) key given an integer seed.
+  
+  ## Examples
+  
+      iex> Nx.Random.key(12)
+      #Nx.Tensor<
+        u32[2]
+        [0, 12]
+      >
+  
+      iex> Nx.Random.key(999999999999)
+      #Nx.Tensor<
+        u32[2]
+        [232, 3567587327]
+      >
+  """
+  defn key(seed) do
+    k1 = Nx.right_shift(seed, 32)
+    k2 = Nx.bitwise_and(seed, 0xFFFFFFFF)
+
+    Nx.stack([k1, k2])
+    |> Nx.as_type({:u, 32})
+  end
+
+  @doc """
+  Splits a PRNG key into `num` new keys by adding a leading axis.
+  
+    ## Examples
+  
+      iex> key = Nx.Random.key(1701)
+      iex> Nx.Random.split(key, 2)
+      #Nx.Tensor<
+        u32[2][2]
+        [
+          [56197195, 1801093307],
+          [961309823, 1704866707]
+        ]
+      >
+  
+      iex> key = Nx.Random.key(999999999999)
+      iex> Nx.Random.split(key, 4)
+      #Nx.Tensor<
+        u32[4][2]
+        [
+          [3959978897, 4079927650],
+          [3769699049, 3585271160],
+          [3182829676, 333122445],
+          [3185556048, 1258545461]
+        ]
+      >
+  """
+  defn split(key, num \\ 2) do
+    assert_key(key)
+
+    impl(key, Nx.iota({num, 2}))
+  end
+
+  defn fold_in(key, data) do
```

Comment:
Should this be documented? Or should it be private?

Reply:
Should be private but I will leave tests commented for now.